[GH-277] Allow selecting labels when creating new issue

utkuufuk opened the pull request:
#### Summary
This PR will allow searching for & selecting labels for the newly created issue in the "Create Issue" dialog.

![image](https://user-images.githubusercontent.com/15212283/89461799-216a9500-d775-11ea-8835-f05c7996d564.png)

---

**NOTE:** The new selector widget will not allow creating new labels on the fly even though the GitHub API allows it. This is for preventing label creation by mistake due to user typos.

#### Ticket Link
Fixes #277

mattermod commented:
Thanks @utkuufuk for the pull request!

 Per the [**CONTRIBUTING.md**](https://github.com/mattermost/platform/blob/master/CONTRIBUTING.md#submitting-a-pull-request) file displayed when you created this pull request, we need to add you to the list of [approved contributors](https://docs.google.com/spreadsheets/d/1NTCeG-iL_VS9bFqtmHSfwETo5f-8MQ7oMDE5IUYJi_Y/pubhtml?gid=0&single=true) for the Mattermost project.

**Please help complete the Mattermost [contribution license agreement](http://www.mattermost.org/mattermost-contributor-agreement/)?**

This is a standard procedure for many open source projects. Your form should be processed within 24 hours and reviewers for your pull request will be able to proceed.

Please let us know if you have any questions.

We are very happy to have you join our growing community! If you're not yet a member, please consider joining our [Contributors community channel](https://pre-release.mattermost.com/core/channels/tickets) to meet other contributors and discuss new opportunities with the core team.

utkuufuk commented on .editorconfig:
```diff
@@ -0,0 +1,35 @@
+# https://editorconfig.org/
+# https://marketplace.visualstudio.com/items?itemName=EditorConfig.EditorConfig
```
I don't know how many people use VS Code, but for those who do, this extension can make life much easier :slightly_smiling_face: 

This is just a suggestion, if you don't like it just let me know and I'll delete this file.

hanzei replied:
Thanks for educating be about this extension :+1: 

https://github.com/mattermost/mattermost-plugin-github/pull/345 will include a standardised `.editorconfig` that all Mattermost plugins use. Would you please remove this file?

utkuufuk commented:
@hanzei Could you please take a look at the current state of the PR and let me know if it aligns with the desired goal overall?

I wanted to check with you first before progressing further since I still don't know your thoughts about https://github.com/mattermost/mattermost-plugin-github/issues/277#issuecomment-664637083

I'd rather create 2 more issues for assignee & milestone selection respectively, where the individual specs can be discussed in more detail. It would also make review & QA simpler. That's just my personal opinion though :slightly_smiling_face:

utkuufuk commented on server/plugin/api.go:
```diff
@@ -965,6 +966,33 @@ func (p *Plugin) getPrByNumber(w http.ResponseWriter, r *http.Request, userID st
+	}
+
+	githubClient := p.githubConnect(*info.Token)
+	result, _, err := githubClient.Search.Labels(context.Background(), repoID, query, &github.SearchOptions{})
```
I couldn't find a way to consume the following endpoint using the `go-github` library:
https://docs.github.com/en/rest/reference/issues#list-labels-for-a-repository

It would make things much simpler if we were able to fetch all available labels instead of searching by query string.

utkuufuk commented on webapp/src/components/github_label_selector/github_label_selector.jsx:
```diff
@@ -0,0 +1,79 @@
+                    className='basic-multi-select'
+                    classNamePrefix='select'
+                    styles={getStyleForReactSelect(this.props.theme)}
+                    onInputChange={this.fetchLabels}
```
I know this isn't ideal, but I wasn't able to come up with a more efficient solution due to https://github.com/mattermost/mattermost-plugin-github/pull/343#discussion_r464065428

utkuufuk commented on webapp/src/components/github_repo_selector/github_repo_selector.jsx:
```diff
@@ -34,7 +34,16 @@ export default class GithubRepoSelector extends PureComponent {
 
     onChange = (name, newValue) => {
-        this.props.onChange(newValue);
+        const newID = this.props.yourRepos.reduce((id, repo) => {
+            return repo.full_name === newValue ? repo.id : id;
+        }, 0);
+
+        // return if there's no match (ideally this should never happen)
+        if (newID === 0) {
+            return;
+        }
+
+        this.props.onChange(newID, newValue);
```
We retrieve labels by repository ID, so we have to save it here along with the name.

utkuufuk commented:
@mickmister As per your suggestion on https://github.com/mattermost/mattermost-plugin-github/issues/277#issuecomment-667587981, I updated the implementation such that we're now using the new `BackendSelector` component to pre-fetch available labels from the server. I had to make some changes to the version in the Jira plugin though; I believe `AsyncSelect` isn't what we're looking for in the current use case, so I replaced it with `ReactSelect`.

(If I'm not mistaken, `AsyncSelect` loads options from the remote source interactively as the user types, which is not what we want.)

@mickmister @hanzei If you think the current implementation is good, I'll proceed with adding assignee & milestone selector widgets using the same approach.

mickmister commented on webapp/src/components/backend_selector.jsx:
```diff
@@ -0,0 +1,137 @@
+    }
+
+    render = () => {
+        this.loadOptions();
```
This currently calls `this.loadOptions` on every render, which doesn't seem like something we want to do. We instead would want to load these in `componentDidMount` since they will only be loaded once. This is taken care of in the `BackendSelector` component from the Jira plugin.

utkuufuk replied:
@mickmister The reason I had to call this on every render was that we have to reload the labels whenever the selected repository changes. 

More specifically, `GithubLabelSelector` re-renders when the selected repository changes, which also causes a re-render for this component. If I don't call `loadOptions` within `render`, our list of labels will become stale. 

That said, I'm definitely open to suggestions on how we can make this more efficient.

mickmister replied:
We can instead use the [componentDidUpdate](https://reactjs.org/docs/react-component.html#componentdidupdate) method, and compare the previous repo value with the current one being passed in. If the value has changed, then we re-fetch the options.

mickmister commented on webapp/src/components/backend_selector.jsx:
```diff
@@ -0,0 +1,137 @@
+
+        return (
+            <Setting {...this.props}>
+                <ReactSelect
+                    required={this.props.required}
+                    isMulti={this.props.isMulti}
+                    name={this.props.name}
+                    onChange={this.onChange}
+                    options={this.state.options}
+                    styles={getStyleForReactSelect(this.props.theme)}
+                />
```
Since five of these props are passed in from the parent, we can shorthand them as `...this.props`
```suggestion
                <ReactSelect
                    {...this.props}
                    styles={getStyleForReactSelect(this.props.theme)}
                />
```

mickmister commented on webapp/src/components/github_label_selector/github_label_selector.jsx:
```diff
@@ -0,0 +1,63 @@
+
+import BackendSelector from 'components/backend_selector';
+
+export default class GithubLabelSelector extends Component {
```
Instead of implementing `shouldComponentUpdate`, we can use React's built-in `PureComponent`

```suggestion
export default class GithubLabelSelector extends React.PureComponent {
```

mickmister commented on webapp/src/components/modals/create_issue/create_issue.jsx:
```diff
@@ -169,9 +171,7 @@ export default class CreateIssueModal extends PureComponent {
-                    <Modal.Title>
-                        {'Create GitHub Issue'}
-                    </Modal.Title>
+                    <Modal.Title>{'Create GitHub Issue'}</Modal.Title>
```
I like all of the style improvements in this PR, but I personally think this one is more readable with the content on its own line as it was before

mickmister commented:
> If I'm not mistaken, AsyncSelect loads options from the remote source interactively as the user types, which is not what we want.

@utkuufuk What if we instead supported a flag in the component that said to only fetch the initial values? Something like:

```jsx
const Select = this.props.initialOnly ? ReactSelect : AsyncSelect;

<Select
    {...this.props}
    styles={getStyleForReactSelect(this.props.theme)}
/>
```

There will likely be other things that need to be done to adjust the props for `AsyncSelect`, but this is the general idea. We should be able to just add the [BackendSelector](https://github.com/mattermost/mattermost-plugin-jira/blob/8a6eb97e010afb9ddbf548112a749a50e60b65a2/webapp/src/components/data_selectors/backend_selector.tsx) to this project and adjust it to support the `initialOnly` prop. Also, it's already written in Typescript, so we would ideally keep it as Typescript.

utkuufuk commented on webapp/.gitignore:
```diff
@@ -1,2 +1,3 @@
 node_modules
 .npminstall
+dist
```
@mickmister Actually this isn't mandatory, as the top-level `.gitignore` takes care of this folder. I added it here too so that tools [like this](https://github.com/ggreer/the_silver_searcher) doesn't show matches from `dist` when run from within the `webapp` directory.

hanzei replied:
I would like to keep `.gitignore` consistent with the other plugin repos. Would you mind reverting these change?

utkuufuk replied:
Sure :+1:

utkuufuk commented:
@mickmister Thank you for reviewing. Here are my opinions on some of your suggestions. Please let me know what you think 🙂 

> What if we instead supported a flag in the component that said to only fetch the initial values?

The thing is, we won't need `AsyncSelect` in the new feature that this PR introduces. (Similarly to labels, assignees and milestones will also be pre-fetched using `ReactSelect`.) Personally, I'm a big fan of deferring complexity until it's absolutely needed, so I'd rather keep it simple for now and extend the component when the need arises.

This is just my personal opinion though. I can definitely implement your suggestion if you feel strongly about it. 🙂

---

> Also, it's already written in Typescript, so we would ideally keep it as Typescript.

I didn't want to convert it to JavaScript too, but TypeScript hasn't been set up for this repository yet. I would have to make a lot of configuration changes to keep it as TypeScript, which would be outside of this PR's scope IMHO. What are your thoughts on creating a separate issue for setting up TypeScript for this repository and converting individual components?

mickmister commented:
@utkuufuk

> The thing is, we won't need AsyncSelect in the new feature that this PR introduces.

You are right, and nothing in this repo needs the server-side autocomplete yet, so we don't need to use `AsyncSelect`. I'm wondering if it makes sense to use it for assignees though. There may be quite a few assignees in a given select box. @hanzei What are your thoughts on pre-fetching all assignees for a given repo, versus using server-side autocomplete search as the user is typing?

The name for the select would ideally have some implication of "these things are attached to an issue/PR, and they are all scoped to a given repo." Mainly because it internally refreshes when the repo changes. Maybe we should name it `IssueAttributeSelector` or something similar.

-----

>  but TypeScript hasn't been set up for this repository yet

Whoops, I was not aware of this. I am currently in the process of sync'ing repositories to include things like introducing typescript, so I will sync this one next.

-----

When the selected repo changes, I think we will want to do the following for each dependent selector:

- Use the `isLoading` prop for the `ReactSelect` component, which will show a loading indicator in the select box
- After the new items are fetched, clear any choices that do not exist in the newly selected repo

utkuufuk commented:
> 'm wondering if it makes sense to use it for assignees though. There may be quite a few assignees in a given select box. @hanzei What are your thoughts on pre-fetching all assignees for a given repo, versus using server-side autocomplete search as the user is typing?

@mickmister @hanzei `ReactSelect` can also filter matching results from the pre-fetched options as the user starts typing. 

I think using `AsyncSelect` would make sense if there were hundreds of possible assignees (or more) for a given repo, and we don't want to fetch a very large amount of data in advance. Correct me if I'm wrong but in our case, I don't think there will be that many possible assignees for a repo most of the time.

mickmister commented:
@utkuufuk Using GitHub's assignee selector with the dev tools network tab open, it looks like they aren't even doing server-side autocomplete while typing. So we should be good using `ReactSelect` :+1:

mickmister commented on webapp/src/components/issue_attribute_selector.jsx:
```diff
@@ -60,13 +61,30 @@ export default class IssueAttributeSelector extends PureComponent {
-        if (!options) {
+    loadOptions = async () => {
+        this.setState({isLoading: true});
+        const options = await this.props.load();
```
We should handle the error case here, by setting the error in this component's state, and setting `isLoading` to false.

mickmister commented on webapp/src/components/github_label_selector/github_label_selector.jsx:
```diff
@@ -0,0 +1,59 @@
+        }).isRequired,
+    };
+
+    handleChange = (items) => {
+        if (!items || items.length === 0) {
+            return;
+        }
+
+        this.props.onChange(items);
+    };
```
Why do we not want to call `this.props.onChange` when there are no values selected? I assume this happens when the user clears the values

mickmister commented on webapp/src/components/github_label_selector/github_label_selector.jsx:
```diff
@@ -0,0 +1,59 @@
+    };
+
+    render() {
+        const loadLabels = async () => {
```
This can be a method on the component, rather than a function declared inside of `render`?

mickmister commented on webapp/src/components/github_label_selector/github_label_selector.jsx:
```diff
@@ -0,0 +1,59 @@
+                return [];
+            }
+
+            await this.props.actions.getLabels(this.props.repo);
+
+            return this.props.labels.map((item) => ({
+                value: item.name,
+                label: item.name,
+            }));
```
Can we use the return value of `getLabels` to construct the options, instead of implicitly using `this.props.labels` after the call is complete?

mickmister commented on webapp/src/components/github_label_selector/github_label_selector.jsx:
```diff
@@ -0,0 +1,59 @@
+                    isMulti={true}
+                    theme={this.props.theme}
+                    onChange={this.handleChange}
+                    load={loadLabels}
```
What do you think about making this more descriptive as `loadOptions`?

```suggestion
                    loadOptions={loadLabels}
```

mickmister commented on webapp/src/components/issue_attribute_selector.jsx:
```diff
@@ -0,0 +1,150 @@
+        }
+
+        if (prevProps.repo !== this.props.repo) {
+            this.loadOptions();
```
Should we check that `this.props.repo` is not empty here?

mickmister commented on webapp/src/components/issue_attribute_selector.jsx:
```diff
@@ -0,0 +1,150 @@
+    }
+
+    onChange = (values) => {
+        this.setState({values});
+
+        if (!values) {
+            this.props.onChange(this.props.isMulti ? [] : '');
+            return;
+        }
```
This is usually an anti-pattern to have values in the parent's state and the child component's state. I recommend removing the `values` property from this component's state, and instead always relying on the parent's `value` prop as the source of truth.

utkuufuk replied:
@mickmister This is a good point. I realized that the `value` prop and some other props were actually unused in this component, so I got rid of those. 

Now there's only one source of truth for the selected values and it's in this component's state. It's a bit simpler to keep this state here as opposed to moving it to the parent's state because otherwise, I'd have to define states for the upcoming assignee & milestone selector components as well.

mickmister replied:
@utkuufuk What do you think about using the prop instead of storing the value in the child component's state? I feel very strongly about only storing the component's value in the parent. It's fine to store the options in the child component, as the parent doesn't need to know about the available choices, but the selected values should only be stored in one place.

mickmister replied:
Currently the selected values are being stored in the parent's `labels` state, and the child component's `values` state. I propose only storing them in the `labels` state, and have the child component rely on the prop coming in. We want to have mostly stateless [controlled components](https://reactjs.org/docs/forms.html#controlled-components) whenever possible.

utkuufuk replied:
@mickmister Ahh you're 100% correct. I completely missed the point that we're storing the labels in `CreateIssueModal` too :smile:

utkuufuk replied:
@mickmister Done: https://github.com/mattermost/mattermost-plugin-github/pull/343/commits/3e3dc2537f9683d398dd6359b8d4f685bce03e2d

mickmister commented on webapp/src/components/issue_attribute_selector.jsx:
```diff
@@ -0,0 +1,150 @@
+            return true;
+        }
+
+        const valid = Boolean(this.props.value && this.props.value.toString().length !== 0);
```
`this.props.value.toString().length` is a clever way to check for presence, but I would like this to be a bit more explicit. The fact that `[].toString() === ''` is not something that's intuitive. Does the following work?

```suggestion
        const valid = Boolean(this.props.value && this.props.value.length);
```

mickmister commented on webapp/src/components/issue_attribute_selector.jsx:
```diff
@@ -0,0 +1,150 @@
+        );
+    };
+
+    render = () => {
```
Not important, but we usually declare the `render` function is this way

```suggestion
    render() {
```

mickmister commented on webapp/src/components/issue_attribute_selector.jsx:
```diff
@@ -0,0 +1,150 @@
+                    className='fa fa-warning'
+                    title='Warning Icon'
+                />
+                <span> {this.state.error.toString()}</span>
```
Any reason why we need the call to `toString()`? If it is a real error we can use the `message` prop, otherwise just use the value of the `error` if it's a string. Although, if we are unsure if it's an error or string, I suppose `toString()` works for both cases.

```suggestion
                <span>{this.state.error.message}</span>
```

mickmister commented on webapp/src/components/modals/create_issue/create_issue.jsx:
```diff
@@ -150,6 +145,13 @@ export default class CreateIssueModal extends PureComponent {
                 />
                 {issueTitleValidationError}
+
+                <GithubLabelSelector
+                    repo={this.state.repoValue}
+                    theme={theme}
+                    onChange={this.handleLabelsChange}
+                />
```
Usually whenever a component is passing an `onChange` prop, there is always an accompanying `value` prop passed as well. This lets the parent hold the state so it's not duplicated in the child component.

utkuufuk replied:
@mickmister See https://github.com/mattermost/mattermost-plugin-github/pull/343#discussion_r465818978

utkuufuk commented:
@mickmister I've addressed all your feedback on your last review, and also implemented assignee & milestone selection. The PR is now complete functionality-wise. Could you please take another look and let me know what you think?

cc @hanzei

hanzei commented on .gitignore:
```diff
@@ -1,3 +1,4 @@
-node_modules
\ No newline at end of file
+node_modules
+.prettierrc.yml
```
What tool does use this file?

utkuufuk replied:
@hanzei The Prettier VS Code extension uses it: https://marketplace.visualstudio.com/items?itemName=esbenp.prettier-vscode

hanzei replied:
In that case I think it should be place in the global `.gitignore` and not the a repository spefic one.

hanzei commented on server/plugin/api.go:
```diff
@@ -979,6 +982,81 @@ func (p *Plugin) getPrByNumber(w http.ResponseWriter, r *http.Request, userID st
+	}
+
+	githubClient := p.githubConnect(*info.Token)
+	result, _, err := githubClient.Issues.ListLabels(context.Background(), owner, repo, &github.ListOptions{})
```
All list API methods from the `github` package must use pagination. Otherwise only the first 30 options will be show. See https://pkg.go.dev/github.com/google/go-github/v31/github?tab=doc#hdr-Pagination for more information and sample code.

hanzei commented on server/plugin/api.go:
```diff
@@ -1094,8 +1175,16 @@ func (p *Plugin) createIssue(w http.ResponseWriter, r *http.Request, userID stri
+	}
+
+	// submitting the request with an invalid milestone ID results in a 422 error
+	// we make sure it's not zero here, because the MM client might have left this field empty
```
Nit pick: What is "the MM client"? Can the documentation be more specific.

utkuufuk replied:
@hanzei I was referring to the front-end HTTP client in this plugin: https://github.com/mattermost/mattermost-plugin-github/blob/master/webapp/src/client/client.js

I could reword "MM client" as "front-end client" unless you have another suggestion?

hanzei replied:
`webapp client` or just `webapp` is a term often used in MM. But `front-end client` is also fine.

hanzei commented on server/plugin/api.go:
```diff
@@ -1180,3 +1271,17 @@ func (p *Plugin) getToken(w http.ResponseWriter, r *http.Request) {
 }
+
+// parseRepo parses the owner & repository name from the repo query parameter
+func parseRepo(repoParam string) (owner, repo string, err error) {
```
Nice helper :+1:

hanzei commented on webapp/src/components/issue_attribute_selector.jsx:
```diff
@@ -0,0 +1,105 @@
+                    styles={getStyleForReactSelect(this.props.theme)}
+                />
+
+                {this.state.error && (
+                    <p className='alert alert-danger'>
+                        <i
+                            className='fa fa-warning'
+                            title='Warning Icon'
+                        />
+                        <span> {this.state.error.message}</span>
+                    </p>
+                )}
```
I tried to test this code by modifying all API endpoints to return 500.

Should an error icon be shown in this case?
![Screenshot from 2020-08-06 11-48-53](https://user-images.githubusercontent.com/16541325/89518178-d71bff00-d7da-11ea-8e6c-cf112ced8df0.png)

utkuufuk replied:
@hanzei Very good catch, I fixed it. This is how it looks now in case of a server error:

![image](https://user-images.githubusercontent.com/15212283/89560901-6b5c8500-d820-11ea-9a04-344bcf017a64.png)

hanzei replied:
Thanks for fixing this issue :+1: 

I've found a minor edge case while testing:
1. Modify all endpoints to return 500 and deploy
3. Open "Create issue dialog"
4. Select a repository
5. Confirm that error message is shown
6. Revert changes to endpoint and deploy to simulate a partial outage
7. Select another repository

Expected:
Error message disappears once new repository is selected. Labels, Assignees and Milestones can be correctly selected.

Observed:
Error message persists once new repository is selected. Labels, Assignees and Milestones can be correctly selected.


This is totally an edge case and not a blocker for this PR. Feel free to ignore it, if it's to hard to fix.

utkuufuk replied:
@hanzei Thanks for pointing that out, I fixed the bug :+1:

hanzei commented on webapp/src/components/github_milestone_selector/github_milestone_selector.jsx:
```diff
@@ -0,0 +1,76 @@
+                <label className='control-label margin-bottom x2'>
+                    {'Milestone'}
+                </label>
+                <IssueAttributeSelector
```
Once a milestone is selected, there is no option to remove it. 
![Screenshot from 2020-08-06 11-53-20](https://user-images.githubusercontent.com/16541325/89518642-79d47d80-d7db-11ea-8158-5d096b3f4ada.png)

For labels and assignees there is a `x` on the right side
![Screenshot from 2020-08-06 11-53-48](https://user-images.githubusercontent.com/16541325/89518710-97094c00-d7db-11ea-9100-cc22bc078e4b.png)

hanzei replied:
The little `x` on the right side of Labels and Assignees is not gone. Was this change intentional? 

![Screenshot from 2020-08-07 09-52-23](https://user-images.githubusercontent.com/16541325/89622901-0b072b00-d894-11ea-90e4-aaf034ff9b65.png)

utkuufuk replied:
@hanzei Why do we want to get rid of the `x` on the right side of labels & assignees? I think it's a good thing that the user is able to de-select a single label/assignee rather than having to clear all of the selected items. 

However, I realized that I removed the "clear all" `x` button on the far right side by mistake, so I'll add it back.

hanzei replied:
Sorry, I had a typo in my comment. The `x` should in fact be there in the RHS.

utkuufuk commented:
> Looking good @utkuufuk! I just have one request regarding how we store the selected milestone.

@mickmister In addition to milestones, I also updated the way we store labels & assignees as per your suggestion: https://github.com/mattermost/mattermost-plugin-github/pull/343/commits/2da037b3cfb785c845a0ee2fd7bdd6b224c8c367

hanzei commented on webapp/src/components/github_milestone_selector/github_milestone_selector.jsx:
```diff
@@ -0,0 +1,74 @@
+                    selection={milestone}
+                    loadOptions={this.loadMilestones}
+                    onChange={this.onChange}
+                />
```
Non blocking idea: 

If no repository is selected, `No Options` will be shown. Is it possible to change to text to `Please select a repository first`
![Screenshot from 2020-08-07 09-54-07](https://user-images.githubusercontent.com/16541325/89622991-2f630780-d894-11ea-9d29-86b293e8a241.png)

hanzei commented:
@larkox Could you also take a look at the webapp code of this PR? It's a fairly large change and I want to make sure nothing got missed.

larkox commented:
I still have to review the code, but regarding UX, in other multiselection on the product we use rounded borders. Can we make them round so it is more coherent with the main product?
![Screenshot from 2020-08-07 11-05-45](https://user-images.githubusercontent.com/1933730/89629477-2bd47e00-d89e-11ea-95e4-77edf69fc3ae.png)
CC @asaadmahmood

larkox commented on webapp/src/components/input.jsx:
```diff
@@ -16,8 +16,8 @@ export default class Input extends PureComponent {
-        addValidate: PropTypes.func.isRequired,
-        removeValidate: PropTypes.func.isRequired,
+        addValidate: PropTypes.func,
+        removeValidate: PropTypes.func,
```
I wonder why this change.

utkuufuk replied:
@larkox Because there are some cases where these props aren't being passed to this component, e.g. in `AttachIssueModal`. This change is actually out of this PR's scope but I was getting warnings on the console so I decided to fix it.

larkox commented on webapp/src/reducers/index.js:
```diff
@@ -102,6 +102,33 @@ function yourRepos(state = [], action) {
     }
 }
 
+function labels(state = [], action) {
```
Not sure if we ever read those states. If we don't, why even implement these as actions?

utkuufuk replied:
@larkox You're correct, we don't need redux for label & assignee options. Here's the fix: https://github.com/mattermost/mattermost-plugin-github/pull/343/commits/a19be4e2e50ac8eb16ae778aa639320da73aa3f3

mickmister commented on webapp/src/components/modals/create_issue/create_issue.jsx:
```diff
@@ -16,6 +19,9 @@ const initialState = {
     issueDescription: '',
+    labels: [],
+    assignees: [],
+    milestone: 0,
```
@utkuufuk Thanks for addressing the strategy of storing the user selected values :+1:

One last request from me. Can we make the default value for milestone `null` instead of `0`? I realize `0` is falsy, but it's not clear that it represents the "empty" value. Setting the initial value to `null`, and subsequently the value we set when user has deselected the milestone, makes it clear that the value is empty.

larkox commented on webapp/src/components/github_milestone_selector/github_milestone_selector.jsx:
```diff
@@ -0,0 +1,74 @@
+        }
+
+        // we have to find the selected milestone from the options in order to insert its number
+        const milestone = this.props.milestones.find((m) => m.title === selection);
```
Can't we just pass the number as the option value?

utkuufuk replied:
@larkox So let's take this milestone as an example: https://github.com/mattermost/mattermost-plugin-github/milestone/7

I think it's better to show the option for this milestone as `v1.2.0` rather than it's number, which is 7.

larkox replied:
I am not sure I am following, and I might not understand completely how React-Select works.
On a dropdown you have a "label" and a "value". The "value" is the "internal" value used for computation. The "label" is the user facing text.
If we set the label to "m.title", the user will see "v1.2.0". If then we set the value to milestone.number, we can work with the value we need (number). That would allow us to not use the redux store here and make the logic way more simple.
Does react-select works different as this?

utkuufuk replied:
@larkox Your suggestion makes sense to me personally. 

However, as per @mickmister's earlier suggestion (https://github.com/mattermost/mattermost-plugin-github/pull/343#discussion_r466608271), we're only storing the milestone number in the parent component state.

So, in the `render` method of `GithubMilestoneSelector`, we have to loop through `this.props.milestones` in order to find the label for the selected milestone so that the label can be rendered in the widget rather than the number.

If we kept track of the milestone label in the parent component state as well as the milestone number, we wouldn't need to keep milestone options in redux state to loop through. But in that case, the way we store labels & assignees in the parent component state would be different than the way we store milestone. (We only store arrays of strings for labels & assignees in the parent component state.)

@larkox @mickmister Please let me know how we want to move forward with this. Personally, I'm fine with either option.

larkox replied:
I get what you are saying, but I would like to see milestone removed from the redux store.

To keep track of the label, we can store the whole milestone, or a simplified version with label and value. That would not conflict with `@mickmister`'s comment, I think, since setting this to null will mean "no selection", and no "truthy" value would be in place. But I am surprised that React-Select does not handle well just passing the value to the `value` prop, as would a normal html select component. Also, I believe that making the simplification of `label == value` is simplifying too much in issue attribute selector. Most of the times you want a human readable value for label, and a machine readable value for value.

@mickmister thoughts?

utkuufuk replied:
@larkox I updated the PR as per your suggestion.

mickmister replied:
When using react-select, it's common to just store the `value` in state, then do something like this in the `render` function. We need to tell it which "option" or "options" are selected.

```js
const selectedOption = this.state.options.find(opt => opt.value === this.state.value);
return (
	<ReactSelect
		options={this.state.options}
		value={selectedOption}
	/>
);
```

mickmister commented:
> I still have to review the code, but regarding UX, in other multiselection on the product we use rounded borders. Can we make them round so it is more coherent with the main product?

@larkox That's an interesting point. The react-select component is used often in plugin modals, so the rounded items style should probably be applied to all plugins if that is the plan of action here.

larkox commented on webapp/src/components/github_milestone_selector/index.js:
```diff
@@ -0,0 +1,23 @@
+});
+
+const mapStateToProps = (state) => ({
+    milestones: state[`plugins-${pluginId}`].milestones,
```
It feels weird we are using the general state for this, specially since we are not using it on other similar places.

utkuufuk replied:
@larkox We don't do this in other places because we don't need this kind of a name-number conversion. For instance, for the label widget, the label name is both what the user sees as an option, and what we pass to the GitHub API. But for milestones, the GitHub API expects the milestone number, not its name. 

Please also see https://github.com/mattermost/mattermost-plugin-github/pull/343#discussion_r467165842

utkuufuk commented on server/plugin/api.go:
```diff
@@ -1120,12 +1236,14 @@ func (p *Plugin) createIssue(w http.ResponseWriter, r *http.Request, userID stri
 	githubClient := p.githubConnect(*info.Token)
 	result, resp, err := githubClient.Issues.Create(context.Background(), owner, repoName, ghIssue)
 	if err != nil {
+		mlog.Error(fmt.Sprintf("Failed to create issue: %v", err.Error()))
```
@larkox Sure, I can use `API.LogError` instead. But just so you know, `mlog` is being used in other places in this file too.

hanzei replied:
I'm inclined to go with consistency here and use `mlog`. There is already https://github.com/mattermost/mattermost-plugin-github/issues/203 filled to capture the effort of using `api.LogX`.

utkuufuk commented:
@larkox @mickmister I addressed most of the comments from the previous round of reviews. Only a few of them remain unresolved, as there are follow-up comments for those either from me or from another developer. Could you please take another look and let me know what you think?

mickmister commented on webapp/src/components/github_milestone_selector.jsx:
```diff
@@ -5,57 +5,42 @@ import React, {PureComponent} from 'react';
-
-        if (options.error) {
+        try {
+            const options = await Client.getMilestones(this.props.repo) || [];
```
We should try to avoid using the `Client` directly in the components. These API calls can be in the `actions.js` file, and passed in through props, even if we do not store the results in redux. @larkox What are your thoughts?

larkox replied:
It is indeed a standard in our projects to do that that way. Probably it will be better that way.

utkuufuk replied:
@mickmister @larkox Done: https://github.com/mattermost/mattermost-plugin-github/pull/343/commits/3ba4df8c3c51abe5e7d9d302ff64e3f36f61647e

mickmister commented on webapp/src/actions/index.js:
```diff
@@ -158,71 +158,35 @@ export function getYourPrsDetails(prList) {
-        const connected = await checkAndHandleNotConnected(data)(dispatch, getState);
-        if (!connected) {
-            return {error: data};
-        }
```
I think we still need to do this step to handle the error case where the user is not connected

mickmister commented on webapp/src/components/github_assignee_selector/index.js:
```diff
@@ -15,4 +15,4 @@ const mapDispatchToProps = (dispatch) => ({
     mapDispatchToProps
-)(GithubMilestoneSelector);
\ No newline at end of file
+)(GithubAssigneeSelector);
```
Nitpick, but do you mind adding a newline to the end of these files?

utkuufuk commented:
> Thanks @utkuufuk! Sorry for the persistence here. I have a request related to the last change made.

@mickmister No problem at all! Thank you for pointing out those issues :slightly_smiling_face:

DHaussermann commented:
@utkuufuk I've done a round of testing on this and it's looking good. 

There is a case I wonder if you're able to handle. When the connected user has access to create issues but does **not** have access to manipulate labels, is it possible to prevent the user from selecting labels in the UI? I'm not sure if you know at the time the repo is selected that there is no permissions on labels?

The current behavior is that the modal will show me the labels and let me select them but when I save the issue, it will have no labels. This could cause confusion and appear to be a bug.

utkuufuk commented:
> The current behavior is that the modal will show me the labels and let me select them but when I save the issue, it will have no labels.

@DHaussermann In the create-issue dialog, I can only select repositories where I'm the owner or I'm a collaborator. In both of those cases, I have permission to assign labels. I don't see this repository, for instance, as an option in the create-issue dialog. I only see my own fork of this repository: https://github.com/utkuufuk/mattermost-plugin-github

Could you help me reproduce the case where you can select a repository that you don't have push access?

cc @hanzei

DHaussermann commented:
Hi @utkuufuk, unfortunately I don't know how to configure the permissions in a way where you can create issues and comments but not add labels. I just happened to have come across a couple of Mattermost repos where this is the case for myself so, I used 1 of those repos to test this case. 

I took a look for this config and I think that when using an Enterprise license on an Organization you can create Teams that get a more granular level of control over the repositories within the org. As explained here in this table https://help.github.jp/enterprise/2.11/user/articles/repository-permission-levels-for-an-organization , it is possible to grant `Read permissions` to a team which allow you to create issues but the permissions for `Apply labels and milestones` is not granted.

Hope this is helpful. Sorry I can't provide a working model. For the time being, it seems I can tests the latest commit to see if the issue is resolved.

utkuufuk commented:
> For the time being, it seems I can tests the latest commit to see if the issue is resolved.

@DHaussermann That would be great. I think there's a good chance that the last commit fixes the bug. I won't have to reproduce this scenario if it does. If not, I'll investigate more.

mickmister commented on webapp/src/components/modals/create_issue/create_issue.jsx:
```diff
@@ -16,7 +16,7 @@ import Input from 'components/input';
     submitting: false,
     error: null,
-    repoValue: '',
+    repo: {},
```
Are there any risks to initializing this as `null`?

```suggestion
    repo: null,
```

utkuufuk replied:
@mickmister Updated the initial repo state as `null`: https://github.com/mattermost/mattermost-plugin-github/pull/343/commits/bfb58bcb4ad6db5c1e170f524dc4c996d9f7c110

mickmister commented on webapp/src/components/modals/create_issue/create_issue.jsx:
```diff
@@ -105,6 +105,37 @@ export default class CreateIssueModal extends PureComponent {
     handleIssueDescriptionChange = (issueDescription) => this.setState({issueDescription});
 
+    renderIssueAttributeSelectors = () => {
+        if (this.state.repo && this.state.repo.permissions && !this.state.repo.permissions.push) {
+            return null;
+        }
```
Does this `permissions.push` determine if they can create labels etc. specifically? I would think that means whether you can create issues/PRs?

utkuufuk replied:
@mickmister It's documented [here](https://docs.github.com/en/rest/reference/issues#create-an-issue) that 
> Any user with pull access to a repository can create an issue.

Also from the same documentation:

![image](https://user-images.githubusercontent.com/15212283/90440903-ad1ee280-e0e0-11ea-9366-957dc31dc583.png)

mickmister commented on webapp/src/components/github_assignee_selector/github_assignee_selector.jsx:
```diff
@@ -0,0 +1,59 @@
+
+export default class GithubAssigneeSelector extends PureComponent {
+    static propTypes = {
+        repo: PropTypes.string.isRequired,
```
Since `state.repo` in the parent is an object, can we change this prop name to `repoName` to avoid confusion? Same with the other similar child components.

utkuufuk replied:
@mickmister Done: https://github.com/mattermost/mattermost-plugin-github/pull/343/commits/e274b08cb89d1c958f9fd2f3ea252d2a19ab26f5

mickmister commented on webapp/src/components/issue_attribute_selector.jsx:
```diff
@@ -0,0 +1,135 @@
+        }
+
+        if (!this.props.selection) {
+            this.props.onChange({});
```
Does this make sense to pass an object here? What assumptions is the parent making based on this?

utkuufuk replied:
@mickmister Good point, I think it'd make more sense to pass `null`, updated: https://github.com/mattermost/mattermost-plugin-github/pull/343/commits/019f192ea44d38ae3f8902616489b1cfe4ad35ad

hanzei commented on webapp/src/components/modals/create_issue/create_issue.jsx:
```diff
@@ -105,6 +105,37 @@ export default class CreateIssueModal extends PureComponent {
     handleIssueDescriptionChange = (issueDescription) => this.setState({issueDescription});
 
+    renderIssueAttributeSelectors = () => {
+        if (this.state.repo && this.state.repo.permissions && !this.state.repo.permissions.push) {
```
Actually, according to https://docs.github.com/en/github/setting-up-and-managing-organizations-and-teams/repository-permission-levels-for-an-organization the required permission is `triage`.

utkuufuk replied:
@hanzei The permissions retrieved from the API call [here](https://github.com/mattermost/mattermost-plugin-github/blob/2061de81547aba5706273b93f8cdf9a87b3ac1cf/server/plugin/api.go#L1011) does not include `triage`. The available permissions are:

* `admin`
* `push`
* `pull`

Also, please take a look at https://github.com/mattermost/mattermost-plugin-github/pull/343#discussion_r471754812

hanzei replied:
You are correct, that API doesn't return `triage`. I'm wondering a bit why that is the case, but we can only work with what the API provides.

hanzei commented:
@utkuufuk One way to reproduce the scenario described by @DHaussermann is by locking the GitHub plugin to an organisation and then grant a member of the org the permission to only read a all repos. In that case, the repos will be shown in the create issue modal, but not labels can be selected.

I leave a minor comment on the required permission.

DHaussermann commented:
Looks like @hanzei still has a code review change pending. I will re-test as needed.

hanzei reviewed (approved):
Thank you very much for this great PR :+1:
